File: src/main/kotlin/ex3/Exam03.java
package ex3;


import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class Exam03 {
    public Exam03(){

        var userList = new ArrayList<User>();

        userList.add(new User("Lee1", 10));
        userList.add(new User("Lee2", 20));
        userList.add(new User("Lee3", 30));

        var lists = Arrays.asList(
                new User("Lee1", 10),
                new User("Lee2", 20),
                new User("Lee3", 30)
        );
        userList.forEach(System.out::println);

        for (int i = 0; i < lists.size(); i++) {
            var userName = lists.get(i).getName();
            System.out.println(userName);
        }

        for(User user: userList){
            System.out.println(user);
        }

        var immutable = Collections.unmodifiableCollection(userList);
        immutable.add(new User("Lee1", 10));

        var imList = List.of("");
        imList.add("ddd");
    }



    public static void main(String[] args){
        new Exam03();
    }
}

class User {
    private String name;
    private int age;

    public User() {
    }

    public User(String name, int age) {
        this.name = name;
        this.age = age;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getAge() {
        return age;
    }

    public void setAge(int age) {
        this.age = age;
    }
}

File: src/main/kotlin/ex3/Exam03.kt
package ex3

fun main(){
    // 가변 컬렉션 (mutable 키워드 시리즈들)
    val userList = mutableListOf<User>()
    userList.add(User("1", 10))
    userList.add(User("2", 20))
    userList.add(User("3", 30))

    // 불변 컬렉션
    val list = listOf<User>(
        User("1", 10),
        User("2", 20)
    )
    
    // list.add -> add 메소드 자체가 없음 (가변과 불변이 명확함)
    // java의 불변 List.of().add()
    // java는 add 메소드가 제공 됨 하지만 실행은 안됨(실수하기 좋음)

    for(user in userList){
        println(user)
    }

    userList.forEach{ it-> println(it) }

    userList.forEachIndexed { index, user ->
        println("index: $index, user: $user")
    }
    
//    상동
//    userList.forEachIndexed(fun (index, user){
//        println("index: $index, user: $user")
//    })

    for ((index, user) in userList.withIndex()){
        println("index: $index, user: $user")
    }

}

// AllAgsConstructor class
class User(var name: String, var age: Int) {
}
